### PR TITLE
Full refactor of this repository

### PR DESCRIPTION
**Pull request by Kreyren**
**DO_NOT_MERGE:** Work in progress..

## **MAINTAINERS**
Review once all checks are completed

#### Required assistance
- [ ] The current build of environments takes longer that i would like -> Theia should be able to install dependencies based on files present in the repository which it should do already? If so provide more info for implementation

## **CONTRIBUTORS**

appending the last layer will be handled explicitely by me since this is a close-source code set for security.

## Checklist (later used as changelog)
- [ ] New "Gitpod docker cores" - Gitpod users has now the option of using `gitpod/<core>` dockerimages that set the mandatory minimums for specified dockerimage
  - [x] Debian
  - [ ] Ubuntu
  - [ ] Gentoo
  - [ ] Exherbo-gnu
  - [ ] Exherbo-musl
  - [ ] Archlinux
  - [ ] Fedora
- [ ] Support for Debian
  - [ ] htop
  - [ ] Openbox (Used in theia desktop)
  - [ ] Full programming language support
      - [ ] Rust
      - [ ] Java
      - [ ] Dotnet
      - [ ] Python
      - [ ] PHP
      - [ ] Nginx
      - [ ] Go
      - [ ] Node.js
      - [ ] Ruby
      - [ ] Haskell
      - [ ] Emmet
      - [ ] Elm
      - [ ] Elixir
      - [ ] Arduino
      - [ ] Free Pascal
      - [ ] Julia
      - [ ] Crystal
      - [ ] Nim
  - [ ] Editors
    - [ ] emacs
    - [ ] vim
    - [ ] nano
  - [ ] Shells
    - [x] Bash
    - [ ] Ion
    - [ ] Powershell
    - [ ] Busybox sh
  - [ ] Linting tools
    - [x] Shellcheck
    - [ ] Pylint
    - [ ] Hadolint
    - [x] valgrind
    - [x] cppcheck
    - [ ] markdownlint
  - [ ] Utilities
    - [ ] Homebrew
    - [x] zip
    - [x] unzip
    - [x] tar
    - [x] Packages for building (build-essentials)
    - [x] less
    - [x] git
    - [X] mirror-grapper implemented
    - [ ] Docker (for development) -- See https://github.com/gitpod-io/gitpod/issues/52
- [ ] Support for Ubuntu
  - [ ] htop
  - [ ] Openbox (Used in theia desktop)
  - [ ] Full programming language support
      - [ ] Rust
      - [ ] Java
      - [ ] Dotnet
      - [ ] Python
      - [ ] PHP
      - [ ] Nginx
      - [ ] Go
      - [ ] Node.js
      - [ ] Ruby
      - [ ] Haskell
      - [ ] Emmet
      - [ ] Elm
      - [ ] Elixir
      - [ ] Arduino
      - [ ] Free Pascal
      - [ ] Julia
      - [ ] Crystal
      - [ ] Nim
  - [ ] Editors
    - [ ] emacs
    - [ ] vim
    - [ ] nano
  - [ ] Shells
    - [ ] Bash
    - [ ] Ion
    - [ ] Powershell
    - [ ] Busybox sh
  - [ ] Linting tools
    - [ ] Shellcheck
    - [ ] Pylint
    - [ ] Hadolint
    - [ ] valgrind
    - [ ] cppcheck
    - [ ] markdownlint
  - [ ] Utilities
    - [ ] Homebrew
    - [ ] zip
    - [ ] unzip
    - [ ] tar
    - [ ] Packages for building (build-essentials)
    - [ ] less
    - [ ] git
    - [ ] Docker (for development) -- See https://github.com/gitpod-io/gitpod/issues/52
 - [ ] Support for Exherbo
   - [ ] exherbo-gnu
    - [ ] htop
    - [ ] Openbox (Used in theia desktop)
    - [ ] Full programming language support
        - [ ] Rust
        - [ ] Java
        - [ ] Dotnet
        - [ ] Python
        - [ ] PHP
        - [ ] Nginx
        - [ ] Go
        - [ ] Node.js
        - [ ] Ruby
        - [ ] Haskell
        - [ ] Emmet
        - [ ] Elm
        - [ ] Elixir
        - [ ] Arduino
        - [ ] Free Pascal
        - [ ] Julia
        - [ ] Crystal
        - [ ] Nim
    - [ ] Editors
      - [ ] emacs
      - [ ] vim
      - [ ] nano
    - [ ] Shells
      - [ ] Bash
      - [ ] Ion
      - [ ] Powershell
      - [ ] Busybox sh
    - [ ] Linting tools
      - [ ] Shellcheck
      - [ ] Pylint
      - [ ] Hadolint
      - [ ] valgrind
      - [ ] cppcheck
      - [ ] markdownlint
    - [ ] Utilities
      - [ ] Homebrew
      - [ ] zip
      - [ ] unzip
      - [ ] tar
      - [ ] Packages for building (build-essentials)
      - [ ] less
      - [ ] git
      - [ ] Docker (for development) -- See https://github.com/gitpod-io/gitpod/issues/52
   - [ ] exherbo-musl
    - [ ] htop
    - [ ] Openbox (Used in theia desktop)
    - [ ] Full programming language support
        - [ ] Rust
        - [ ] Java
        - [ ] Dotnet
        - [ ] Python
        - [ ] PHP
        - [ ] Nginx
        - [ ] Go
        - [ ] Node.js
        - [ ] Ruby
        - [ ] Haskell
        - [ ] Emmet
        - [ ] Elm
        - [ ] Elixir
        - [ ] Arduino
        - [ ] Free Pascal
        - [ ] Julia
        - [ ] Crystal
        - [ ] Nim
    - [ ] Editors
      - [ ] emacs
      - [ ] vim
      - [ ] nano
    - [ ] Shells
      - [ ] Bash
      - [ ] Ion
      - [ ] Powershell
      - [ ] Busybox sh
    - [ ] Linting tools
      - [ ] Shellcheck
      - [ ] Pylint
      - [ ] Hadolint
      - [ ] valgrind
      - [ ] cppcheck
      - [ ] markdownlint
    - [ ] Utilities
      - [ ] Homebrew
      - [ ] zip
      - [ ] unzip
      - [ ] tar
      - [ ] Packages for building (build-essentials)
      - [ ] less
      - [ ] git
      - [ ] Docker (for development) -- See https://github.com/gitpod-io/gitpod/issues/52
- [ ] Support for Gentoo
  - [ ] htop
  - [ ] Openbox (Used in theia desktop)
  - [ ] Full programming language support
      - [ ] Rust
      - [ ] Java
      - [ ] Dotnet
      - [ ] Python
      - [ ] PHP
      - [ ] Nginx
      - [ ] Go
      - [ ] Node.js
      - [ ] Ruby
      - [ ] Haskell
      - [ ] Emmet
      - [ ] Elm
      - [ ] Elixir
      - [ ] Arduino
      - [ ] Free Pascal
      - [ ] Julia
      - [ ] Crystal
      - [ ] Nim
  - [ ] Editors
    - [ ] emacs
    - [ ] vim
    - [ ] nano
  - [ ] Shells
    - [ ] Bash
    - [ ] Ion
    - [ ] Powershell
    - [ ] Busybox sh
  - [ ] Linting tools
    - [ ] Shellcheck
    - [ ] Pylint
    - [ ] Hadolint
    - [ ] valgrind
    - [ ] cppcheck
    - [ ] markdownlint
  - [ ] Utilities
    - [ ] Homebrew
    - [ ] zip
    - [ ] unzip
    - [ ] tar
    - [ ] Packages for building (build-essentials)
    - [ ] less
    - [ ] git
    - [ ] Docker (for development) -- See https://github.com/gitpod-io/gitpod/issues/52
- [ ] Support for Archlinux
  - [ ] htop
  - [ ] Openbox (Used in theia desktop)
  - [ ] Full programming language support
      - [ ] Rust
      - [ ] Java
      - [ ] Dotnet
      - [ ] Python
      - [ ] PHP
      - [ ] Nginx
      - [ ] Go
      - [ ] Node.js
      - [ ] Ruby
      - [ ] Haskell
      - [ ] Emmet
      - [ ] Elm
      - [ ] Elixir
      - [ ] Arduino
      - [ ] Free Pascal
      - [ ] Julia
      - [ ] Crystal
      - [ ] Nim
  - [ ] Editors
    - [ ] emacs
    - [ ] vim
    - [ ] nano
  - [ ] Shells
    - [ ] Bash
    - [ ] Ion
    - [ ] Powershell
    - [ ] Busybox sh
  - [ ] Linting tools
    - [ ] Shellcheck
    - [ ] Pylint
    - [ ] Hadolint
    - [ ] valgrind
    - [ ] cppcheck
    - [ ] markdownlint
  - [ ] Utilities
    - [ ] Homebrew
    - [ ] zip
    - [ ] unzip
    - [ ] tar
    - [ ] Packages for building (build-essentials)
    - [ ] less
    - [ ] git
    - [ ] Docker (for development) -- See https://github.com/gitpod-io/gitpod/issues/52
- [x] Support for other Linux distributions on demand
- [ ] Support for other kernels using Vagrant (on hold, partially implemented and available for contrib)
- [ ] Gitpod layers now support multiple distributions allowing them to be built
- [ ] Various code quality fixes
  - [ ] CircleCI frontend has been rewritten to comply with code quality standard
  - [x] Fixed code quality issues in gitpod-layer
- [ ] New features
  - [ ] End-users can now use `shell` property in their `gitpod.yml` files to use following shells:
    - [ ] Bash
    - [ ] Shell
    - [ ] Redox's ion-shell
    - [ ] Powershell (Using powershell for linux)
  - [ ] End-users can now set PS1 per repository
  - [ ] End-users can now use `package` property in `gitpod.yml` to install various packages without the need to use dockerfile
 - [ ] End-users can now use apt without root
 - [ ] The terminal now speaks multiple languages based on LANG set by the end-user
 - [ ] Thealer! New package management system designed for gitpod
- [ ] Repository management
  - [ ] CI
    - [ ] shell
    - [ ] bash
    - [ ] Assign labels based on file path
  - [ ] CD
    - [ ] Deploy docker images automatically to dockerhub (hub.docker.com)
    - [ ] Deploy valgrant files directly in vagrantup (vagrantup.com)
- [ ] Vagrant frontend support
   - [ ] New "Gitpod Vagrant cores" 
    - [ ] Debian
    - [ ] Ubuntu
    - [ ] Gentoo
    - [ ] Exherbo-gnu
    - [ ] Exherbo-musl
    - [ ] Archlinux
    - [ ] Fedora

---

### References

- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1258
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1217
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1211
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1191
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1144
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1143
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1141
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1095
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1265
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1251
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1095
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/1041
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/845
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/828
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/812
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/805
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/761
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/632
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/474
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/448
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/283
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/240
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/116
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/92
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/84
- [ ] Attempts-to-resolve: https://github.com/gitpod-io/gitpod/issues/30

- [ ] fixes: https://github.com/gitpod-io/gitpod/issues/813
- [ ] fixes: https://github.com/gitpod-io/gitpod/issues/1252
- [ ] fixes: https://github.com/gitpod-io/gitpod/issues/22
- [ ] fixes: https://github.com/gitpod-io/gitpod/issues/17

- [ ] may-resolve: https://github.com/gitpod-io/gitpod/issues/1337
- [ ] may-resolve: https://github.com/gitpod-io/gitpod/issues/1295
- [ ] may-resolve: https://github.com/gitpod-io/gitpod/issues/729
- [ ] may-resolve: https://github.com/gitpod-io/gitpod/issues/80
- [ ] may-resolve: https://github.com/gitpod-io/gitpod/issues/52

- [ ] is-expected-to-help-with: https://github.com/gitpod-io/gitpod/issues/1046
- [ ] is-expected-to-help-with: https://github.com/gitpod-io/gitpod/issues?q=is%3Aissue+is%3Aopen+label%3Aoperations

This follows up on my contribution in gitpod-layer 

### Notes
Images can be tested using 
```
gp preview yolo && make debian && git add * && git commit -m "$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c10)" && git push && gp preview https://gitpod.io/#https://github.com/gitpod-io/workspace-images/pull/204
```

### CODE REVIEW

**Comment by Kreyren** [comment]
NOTICE: The other "cores" does not have the abstract implemented yet, requesting more info for the abstract (See in-code docummentation in case i forgot something) and method to test the images to ensure that it works as expected.

File: .circleci/build_image.sh [hunk 1]
```diff
@@ -1,4 +1,7 @@
 #!/bin/sh
+
+# FIXME(Kreyren): Oh gawd who wrote this, Abstract required for proper implementation
```
Comment (Kreyren) [inline .circleci/build_image.sh:3]: Also i need abstract for the CircleCI to fix code quality and hopefully improve the runtime.

**Comment by Kreyren** [comment]
This should also permanently resolve issues alike https://github.com/gitpod-io/workspace-images/pull/204 and other security and maintainance issues associated with ubuntu.

**Comment by Kreyren** [comment]
Maitainers: No need to review i figured out reliable way to test this commit will ping you when it's finished~

**Comment by Kreyren** [comment]
NOTICE: Shellcheck 0.5.0 is not usable, see false trigger:

image

Code:

```bash
case "$DISTRO" in
	debian)
		case "$expectedShell" in
			bash) apt install -y bash-completion || die 1 "Unable to install bash-completion" ;;
			sh|dash) true ;; # nothing else to do here
			powershell)	apt install -y powershell || die 1 "Unable to install powershell" ;;
			*) die 255 "Unexpected shell in variable expectedShell with value '$expectedShell' has been provided to shellConfig script"
		esac
	;;
	*) die 255 "Unexpected distribution '$DISTRO' has been parsed in shellConfig script"
esac
```

**Comment by Kreyren** [comment]
@jankeromnes @svenefftinge If you want something linux specific then add it in the checklist i will implement it.

**Comment by jankeromnes** [comment]
Hi @Kreyren, please try to avoid sending as many commits to this Pull Request. Many of us receive a notification for every commit, and it also triggers a new full CircleCI workflow, which is quite resource-intensive.

**Comment by Kreyren** [comment]
> Hi @Kreyren, please try to avoid sending as many commits to this Pull Request. Many of us receive a notification for every commit, and it also triggers a new full CircleCI workflow, which is quite resource-intensive.

I though that you won't be getting notifications for draft merge requests?

Noted, closing will reopen once finished.

**Comment by Kreyren** [comment]
Development moved on https://github.com/Kreyren/workspace-images/pull/1

**Pull request closed without merging**